lxd/device/gpu_sriov.go

```diff
@@ -0,0 +1,297 @@
+package device
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/device/pci"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/resources"
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/shared"
+)
+
+type gpuSRIOV struct {
+	deviceCommon
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *gpuSRIOV) validateConfig(instConf instance.ConfigReader) error {
+	if !instanceSupported(instConf.Type(), instancetype.VM) {
+		return ErrUnsupportedDevType
+	}
+
+	requiredFields := []string{}
+
+	optionalFields := []string{
+		"vendorid",
+		"productid",
+		"id",
+		"pci",
+	}
+	err := d.config.Validate(gpuValidationRules(requiredFields, optionalFields))
+	if err != nil {
+		return err
+	}
+
+	if d.config["pci"] != "" {
+		for _, field := range []string{"id", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "pci" is set`, field)
+			}
+		}
+	}
+
+	if d.config["id"] != "" {
+		for _, field := range []string{"pci", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "id" is set`, field)
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateEnvironment checks the runtime environment for correctness.
+func (d *gpuSRIOV) validateEnvironment() error {
+	if d.config["pci"] != "" && !shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s", d.config["pci"])) {
```

Comment:
@monstermunchkin should we be validating `pci` field as lower case to avoid surprises if the user specifies the letters in upper case and cant find the device because its listed in lower case in `/sys/bus/pci/devices`? Or perhaps converting the `pci` value to lower case first?

Reply:
@tomponline I'd convert it before validating it. But we would need to do the same for `vendorid` and `productid`, wouldn't we?

Reply:
@monstermunchkin yes probably as we'd be using them for comparison too (there's probably other places in the code base for PCI devices that would benefit from this). Just a usability thing really. I'm not sure if PCI devices are expected to always be specified as lower case (in which case we could reject upper case letters) or if like MAC addresses, they can be specified either way.

Reply:
I think we'll want to be reasonably lax on what the user can provide, in theory we could support:
 - 0000:XX:XX.X (kernel syntax)
 - XX:XX.X (lspci syntax, assumes 0000: prefix)

And in both upper and lower cases for the letters. We'd then match that to an actual PCI device in `validateEnvironment` and fail if not found.

Reply:
> And in both upper and lower cases for the letters.

@stgraber We require `vendorid` and `productid` to be lower case. Why not do the same for PCI devices?

Reply:
I guess we can do that, it looks like all the usual suspects give it lowercase so that should be fine.